File: src/SoliBee.Core/Models/PokerHandRank.cs
namespace SoliBee.Core.Models;

public enum PokerHandRank
{
    HighCard      = 0,
    OnePair       = 1,
    TwoPair       = 2,
    ThreeOfAKind  = 3,
    Straight      = 4,
    Flush         = 5,
    FullHouse     = 6,
    FourOfAKind   = 7,
    StraightFlush = 8,
    RoyalFlush    = 9,
    FiveOfAKind   = 10,
}

public sealed record PokerHandResult(
    PokerHandRank Rank,
    int PairRank = 0,
    int QuadRank = 0);

File: src/Models/PokerHandRank.swift
import Foundation

public enum PokerHandRank: Int, Comparable, CaseIterable {
    case highCard = 0
    case onePair
    case twoPair
    case threeOfAKind
    case straight
    case flush
    case fullHouse
    case fourOfAKind
    case straightFlush
    case royalFlush

    public static func < (lhs: Self, rhs: Self) -> Bool { lhs.rawValue < rhs.rawValue }

    public var displayName: String {
        switch self {
        case .highCard:     return "High Card"
        case .onePair:      return "One Pair"
        case .twoPair:      return "Two Pair"
        case .threeOfAKind: return "Three of a Kind"
        case .straight:     return "Straight"
        case .flush:        return "Flush"
        case .fullHouse:    return "Full House"
        case .fourOfAKind:  return "Four of a Kind"
        case .straightFlush: return "Straight Flush"
        case .royalFlush:   return "Royal Flush"
        }
    }
}

public struct PokerHandResult: Comparable {
    public let rank: PokerHandRank
    public let kickers: [Int]   // up to 5 card ranks, descending

    public init(rank: PokerHandRank, kickers: [Int]) {
        self.rank = rank
        self.kickers = kickers
    }

    public static func < (lhs: Self, rhs: Self) -> Bool {
        if lhs.rank != rhs.rank { return lhs.rank < rhs.rank }
        return lhs.kickers.lexicographicallyPrecedes(rhs.kickers)
    }

    public static func == (lhs: Self, rhs: Self) -> Bool {
        lhs.rank == rhs.rank && lhs.kickers == rhs.kickers
    }
}
